src/spec.ts

```diff
@@ -319,9 +319,42 @@ export function alwaysNoOcclusion(spec: ExtendedUnitSpec): boolean {
   return vlEncoding.isAggregate(spec.encoding);
 }
 
-export function fieldDefs(spec: ExtendedUnitSpec): FieldDef[] {
+function hasFieldDef(fieldDefs: FieldDef[], fieldDef: FieldDef): boolean {
+  for (let i = 0; i < fieldDefs.length; i++) {
+    if (identical(fieldDefs[i], fieldDef)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+function fieldDefsOfAllLayers(spec: LayerSpec): FieldDef[] {
+  let fieldDefs: FieldDef[] = [];
+  for (let i = 0; i < spec.layers.length; i++) {
+    let fieldDefsInALayer = vlEncoding.fieldDefs(spec.layers[i].encoding);
+    for (let j = 0; j < fieldDefsInALayer.length; j++) {
+      if (!hasFieldDef(fieldDefs, fieldDefsInALayer[j])) {
+        fieldDefs.push(fieldDefsInALayer[j]);
+      }
+    }
+  }
+  return fieldDefs;
+}
+
+export function fieldDefs(spec: ExtendedUnitSpec | LayerSpec | FacetSpec): FieldDef[] {
   // TODO: refactor this once we have composition
-  return vlEncoding.fieldDefs(spec.encoding);
+  if (isLayerSpec(spec)) {
+    return fieldDefsOfAllLayers(spec);
+  } else if (isFacetSpec(spec)) {
+    let facet = spec.spec;
```

Comment:
this chunk for facet looks totally random.   why is `facet = spec.spec`   spec.facet isn't considered at all?